src/js/harness.js

```diff
@@ -1,69 +1,424 @@
+    }
+
+    var command = fluid.stringTemplate(
+        commandTemplate,
+        that.options
+    );
+    exec(command, callback);
+};
+
+/**
+ *
+ * Run a command and resolve/reject a promise when the results are known..
+ *
+ * @param {Object} that - The component itself.
+ * @param {String} commandTemplate - A template representing the command to be run.  Along with `that.options`, will be passed to `fluid.stringTemplate`.
+ * @param {String} [message] - An optional message that will be logged if present.
+ * @return {Promise} A `fluid.promise` instance that will be resolved when the command completes succesfully or rejected if there's an error.
+ *
+ */
+gpii.pouch.harness.runCommandAsPromise = function (that, commandTemplate, message) {
+    var commandPromise = fluid.promise();
+
+    gpii.pouch.harness.runCommand(that, commandTemplate, function (error) {
+        if (error) {
+            commandPromise.reject(error);
+        }
+        else {
+            commandPromise.resolve();
+        }
+    }, message);
+
+    return commandPromise;
+};
+
+/**
+ *
+ * Start monitoring the health of the associated docker container so that we can shut ourselves down if it fails to
+ * start up or is stopped externally using a docker command.
+ *
+ * @param {Object} that - The component itself.
+ *
+ */
+gpii.pouch.harness.startMonitoring = function (that) {
+    that.monitorInterval = setInterval(
+        gpii.pouch.harness.runCommand,
+        that.options.containerMonitoringInterval,
+        that, "docker ps -a --filter name=%containerName --format \"{{.Status}}\"", function (error, stdout) {
+            fluid.log(fluid.logLevel.TRACE, "health check.");
+            // If we can't verify that our container is up, exit.
+            if (error || stdout.indexOf("Up") !== 0) {
+                clearInterval(that.monitorInterval);
+                that.destroy();
             }
+            // Otherwise, keep running, i.e. don't let the node process exit.
         }
-    },
-    components: {
-        express: {
-            type: "gpii.express",
-            options: {
-                "port" : "{harness}.options.port",
-                listeners: {
-                    onStarted: "{harness}.events.expressStarted.fire"
-                },
-                components: {
-                    expressPouch: {
-                        type: "gpii.pouch.express",
-                        options: {
-                            path: "/",
-                            listeners: {
-                                onStarted: "{harness}.events.pouchStarted.fire"
-                            }
-                        }
+    );
+};
+
+/**
+ *
+ * Create the required databases and populate them with data.
+ *
+ * @param {Object} that - The component itself.
+ * @return {Promise} - A `fluid.promise` instance that will be resolved when the data is loaded or rejected if an error occurs.
+ *
+ */
+gpii.pouch.harness.loadData = function (that) {
```

Comment:
This function is too huge to survive as a monolith - please factor appropriately